### ISSUE TITLE
IDE0059 introduces unnecessary local

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#38044]
**Version Used**: 
3.3.0-beta3-19413-06+ac06df1bffb7b7fd0e5bf63cc91921af76b21e03

**Steps to Reproduce**:
```C#
using System;
using System.Runtime.InteropServices;

class Program
{
    static void Main() => Console.WriteLine(GetTopOfMemory());

    private static ulong GetTopOfMemory()
    {
        var info = new SYSTEM_INFO();
        GetSystemInfo(out info);
        return (ulong)info.lpMaximumApplicationAddress;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct SYSTEM_INFO
    {
        internal ushort wProcessorArchitecture;
        internal ushort wReserved;
        internal int dwPageSize;
        internal IntPtr lpMinimumApplicationAddress;
        internal IntPtr lpMaximumApplicationAddress;
        internal IntPtr dwActiveProcessorMask;
        internal int dwNumberOfProcessors;
        internal int dwProcessorType;
        internal int dwAllocationGranularity;
        internal short wProcessorLevel;
        internal short wProcessorRevision;
    }

    [DllImport("kernel32")]
    internal static extern void GetSystemInfo(out SYSTEM_INFO lpSystemInfo);
}
```
Execute IDE0059 on the `var info = new SYSTEM_INFO();` line.

**Expected Behavior**:
Either:
```C#
SYSTEM_INFO info;
GetSystemInfo(out info);
return (ulong)info.lpMaximumApplicationAddress;
```
or
```C#
GetSystemInfo(out SYSTEM_INFO info);
return (ulong)info.lpMaximumApplicationAddress;
```
or IDE0059 isn't offered at all.

**Actual Behavior**:
```C#
_ = new SYSTEM_INFO();
SYSTEM_INFO info;
GetSystemInfo(out info);
return (ulong)info.lpMaximumApplicationAddress;
```
which makes no sense, adding the `_ = new SYSTEM_INFO();`.

**Comment by sharwell** [comment]
Technically the code fix is currently correct for the majority of related cases. The expected behavior is only a strict preservation of behavior if the `SYSTEM_INFO` constructor does not have any observable side effects. In this case we know that to be true because it's equivalent to `default(SYSTEM_INFO)`.

**Comment by stephentoub** [comment]
> Technically the code fix is currently correct for the majority of related cases.

The SYSTEM_INFO struct here is defined in the same assembly, in C#.  It can't possibly have a custom parameterless ctor, and even if it could, the rule can see that it doesn't.

And if there were such a concern, then the rule shouldn't be firing at all, because there's no benefit to:
```C#
_ = new SYSTEM_INFO();
SYSTEM_INFO info;
Foo(out info);
```
over
```C#
SYSTEM_INFO info = new SYSTEM_INFO();
Foo(out info);
```
I can't think of a single situation where I'd want such a refactoring applied.

**Comment by mavasani** [comment]
Yes, the analyzer/code fix here is conservative in presence of expressions that it [does not understand as having no side effects](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs,567). As Sam suggested, we can potentially make the analyzer/fixer more aggressive here for invocations to parameterless struct ctors, which will make this scenario lead to removal of the statement completely.

However, in general the core intent of the analyzer/fixer pair here (especially in context of a FixAll operation) is to draw user's attention onto redundant assignments and the user needs to decide if the value computation is also redundant (i.e. whole statement can be removed) OR computation has side effects whose value is explicitly discarded at this particular call site. Taking the aggressive approach might be potentially risky, and change semantics. Though we can consider offering 2 code fixes here, one which is current conservative and another which is more aggressive and removes the entire assignment. Alternatively, we can also define or enhance the existing editorconfig option for this analyzer to be more aggressive.
